internal/clients/team1/foraging.go

```diff
@@ -0,0 +1,269 @@
+	}
+}
+
+func (c client) constantForage(resourcePercent float64) shared.ForageDecision {
+	decision := shared.ForageDecision{
+		Type:         shared.DeerForageType,
+		Contribution: 0.2 * c.gameState().ClientInfo.Resources,
+	}
+
+	c.Logf("[Forage decision]: constant (%v)", decision)
+	return decision
+}
+
+func outcomeRegression(history []ForageOutcome) regression.Regression {
+	r := new(regression.Regression)
+	r.SetObserved("Team1 reward")
+	r.SetVar(0, "Team1 contribution")
+
+	for _, outcome := range history {
+		r.Train(regression.DataPoint(float64(outcome.revenue), []float64{float64(outcome.contribution)}))
+	}
+
+	r.AddCross(regression.PowCross(0, 2))
+	r.Run()
+
+	return *r
+}
+
+func (c *client) regressionOptimalContribution(r regression.Regression) shared.Resources {
+	if r.Coeff(2) > 0 {
+		// Curves up, mo money is mo money
+		return 0.4 * c.gameState().ClientInfo.Resources
+	}
+
+	// Curves down, mo money is not always mo money
+	// c.Logf("%v x^2 + %v x + %v", r.Coeff(2), r.Coeff(1), r.Coeff(0))
+	// c.Logf("%v / %v = %v ", -r.Coeff(1), 2*r.Coeff(2), shared.Resources(-r.Coeff(1)/(2*r.Coeff(2))))
+	return shared.Resources(-r.Coeff(1) / (2 * r.Coeff(2)))
+}
+
+func (c client) desperateForage() shared.ForageDecision {
+	forageType := bestROIForageType(c.forageHistory)
+	if forageType == shared.ForageType(-1) {
+		forageType = shared.DeerForageType
+	}
+
+	return shared.ForageDecision{
+		Type:         forageType,
+		Contribution: c.gameState().ClientInfo.Resources,
+	}
+}
+
+func (c *client) DecideForage() (shared.ForageDecision, error) {
+	if c.forageHistorySize() < c.config.randomForageTurns {
+		c.Logf("[Forage decision]: random")
+		return c.randomForage(), nil
+	} else if c.emotionalState() == Desperate {
+		c.Logf("[Forage decision]: desperate")
+		return c.desperateForage(), nil
+	} else if len(c.livingClients()) > 1 {
+		c.Logf("[Forage decision]: flip")
+		return c.flipForage(), nil
+	} else {
+		c.Logf("[Forage decision]: constant")
+		return c.constantForage(0.2), nil
+	}
+}
+
+func (c *client) ForageUpdate(forageDecision shared.ForageDecision, revenue shared.Resources) {
+	c.forageHistory[forageDecision.Type] = append(c.forageHistory[forageDecision.Type], ForageOutcome{
+		contribution: forageDecision.Contribution,
+		revenue:      revenue,
+		turn:         c.gameState().Turn,
+		participant:  c.GetID(),
+	})
+
+	c.Logf(
+		"[Forage result]: %v(%05.3f) | Expectation: %+05.3f | Reward: %+05.3f | Error: %.0f%%",
+		forageDecision.Type,
+		forageDecision.Contribution,
+		c.expectedForageReward,
+		revenue,
+		((c.expectedForageReward-revenue)/revenue)*100,
+	)
+}
+
+/************************/
+/***  Forage Helpers  ***/
+/************************/
+
+// bestROIForageType finds the ForageType that has the best average ROI. If all
+// forageTypes have negative returns then it returns shared.ForageType(-1)
+func bestROIForageType(forageHistory ForageHistory) shared.ForageType {
+	bestForageType := shared.ForageType(-1)
```

Comment:
not a fan of this, let's open an issue/pr for a new ForageType called `NoForage`?

or, change the return signature of the func to `(shared.ForageType, participating bool)`

Reply:
(lowpri nitpick)

Reply:
I've already suggested this in #143, but JamesTev said that not foraging should just be 0 resources.

Reply:
Should I just leave it as-is?